### PR TITLE
fix count_wildcard_rule when alias in subquery

### LINKED ISSUE #6447: Cannot have column named "COUNT(*)"
### Describe the bug

After the changes in https://github.com/apache/arrow-datafusion/pull/6010, it's no longer possible to have a column named "COUNT(*)" (even with quotes).  A column with this name is renamed to "COUNT(UInt8(1))".

### To Reproduce

Before #6010:

```
❯ with tbl as (SELECT 0 as "COUNT(*)") SELECT * from tbl;
+----------+
| COUNT(*) |
+----------+
| 0        |
+----------+
```

After #6010

```
❯ with tbl as (SELECT 0 as "COUNT(*)") SELECT * from tbl;
+-----------------+
| COUNT(UInt8(1)) |
+-----------------+
| 0               |
+-----------------+
```

### Expected behavior

I expect it to be possible to have a column named "COUNT(*)" without it being automatically renamed.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jiangzhx**
# Which issue does this PR close?

Closes #6447 .

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by jiangzhx** [comment]
@jackwener hope you have time for review.

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 1]
```diff
@@ -97,19 +98,47 @@ fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
...
+                        };
+                        (expr.clone().rewrite(&mut rewriter).unwrap(), new_field)
+                    })
+                    .unzip();
+            let new_schema = DFSchemaRef::new(
+                DFSchema::new_with_metadata(
+                    new_fields,
+                    projection.schema.metadata().clone(),
+                )
+                .unwrap(),
```
Comment (jackwener) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:138]:
```suggestion
                .unwrap(),
```
```suggestion
                )?,
```

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 2]
```diff
@@ -97,19 +98,47 @@ fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
...
+                                }
+                                DFField::new(
+                                    field.qualifier().cloned(),
+                                    &name,
+                                    field.data_type().clone(),
+                                    field.is_nullable(),
+                                )
+                            }
+                        };
+                        (expr.clone().rewrite(&mut rewriter).unwrap(), new_field)
```
Comment (jackwener) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:130]: `unwrap()` isn't a good choice

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 3]
```diff
@@ -97,19 +98,47 @@ fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
...
+                                    );
+                                }
+                                DFField::new(
+                                    field.qualifier().cloned(),
+                                    &name,
+                                    field.data_type().clone(),
+                                    field.is_nullable(),
+                                )
+                            }
+                        };
```
Comment (jackwener) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:129]: It isn't a good method to resolve this issue.
IMO, it looks like a `hack` patch to fix issue.

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs (hunk 2 above)
Comment (jiangzhx) [reply datafusion/optimizer/src/analyzer/count_wildcard_rule.rs]: thanks,fixed.

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 4]
```diff
@@ -97,7 +97,12 @@ fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
                 fetch,
             })))
         }
-        LogicalPlan::Projection(projection) => {
+        LogicalPlan::Projection(projection)
+            if matches!(
+                projection.input.as_ref(),
+                LogicalPlan::Aggregate(..) | LogicalPlan::Window(..)
+            ) =>
```
[the remark this answers is not in the document]
Comment (jiangzhx) [reply datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:105]: fixed,Thank you for your review.

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs (hunk 3 above)
Comment (jiangzhx) [reply datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:129]: Yes, this piece of code is indeed difficult to understand and seems strange. However, I haven't found a good solution for it yet.

thanks for your help.

Comment (jiangzhx) [reply datafusion/optimizer/src/analyzer/count_wildcard_rule.rs]: i did some change.hope you have time for review.
thanks again.

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 5]
```diff
@@ -98,19 +98,35 @@ fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
...
-                .collect::<Result<Vec<_>>>()?;
-            Ok(Transformed::Yes(LogicalPlan::Projection(
-                Projection::try_new_with_schema(
-                    projection_expr,
-                    projection.input,
-                    // rewrite_schema(projection.schema.clone()),
-                    rewrite_schema(&projection.schema),
-                )?,
-            )))
+            let input = projection
```
Comment (jackwener) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:101]:
```suggestion
            // resolve issue: https://github.com/apache/arrow-datafusion/issues/6447
            // TODO: In the future, we can systematically solve the problem about `name changes` and then revert these code and other code to resolve this kind of problem.
            let input = projection
```

**Comment by jackwener** [comment]
This is a more complicated problem. After discussing with @jiangzhx , I will solve this problem later

**Pull request closed without merging**